blog/content/post/2020-11-30-coreboot_for_power.md: new post (#353)
3mdeb/news-and-ideas: blog/content/post/2020-11-30-coreboot_for_power.md

---
title: coreboot port for OpenPOWER - why bother?
abstract: 'You may have heard by now that we are working on coreboot port for
          Talos II. OpenPOWER already has, nomen omen, open source firmware, so
          one may ask why bother? We will try to answer that question.'
cover: /covers/coreboot-logo.svg
author: krystian.hebel
layout: post
published: true
date: 2020-11-30
archives: "2020"

tags:
  - firmware
  - coreboot
  - openpower
categories:
  - Firmware

---

> You may have heard by now that we are working on coreboot port for [Talos II](https://raptorcs.com/TALOSII/).

Reviewer: not relevant here, but jumping in and clarifying the confusion with a link to this post would totally make sense: https://twitter.com/rozendantz/status/1336113596837720065

OpenPOWER already has, nomen omen, open source firmware, so one may ask why

> bother? In this blog post we will try to answer that question.

Reviewer 2: @KBanecka I would add couple references here OpenPOWER Summit, OSFC, maybe some tweet.

Reviewer 3: nomen omen? never heard this phrase before.

Reviewer 4: It should rather be `nomen atque omen` or `nomen est omen`. Nomen omen is an abbreviation used often in polish

## Short introduction to OpenPOWER boot process and its components

Talos II is a server platform managed by a slightly customised (mainly to handle
specific dual-CPU boot synchronisation quirks) [OpenBMC fork](https://git.raptorcs.com/git/talos-openbmc/).
BMC signals the main (host) platform to start booting.

Next comes the [Self-Boot Engine](https://wiki.raptorcs.com/wiki/Self-Boot_Engine).

> This term is actually used both for a chip (which is part of main CPU, it is a

Reviewer 5: ```suggestion
This term is actually used both for a chip (which is part of main CPU – a
```

reduced PowerPC core) and its [firmware](https://git.raptorcs.com/git/talos-sbe/).
SBE's tasks are initialisation of main CPU cores and loading and staring the

> next component called Hostboot.

Reviewer 6: Technically there is another piece in between.  The SBE code loads up the Hostboot Bootloader first.  This piece of code is responsible for loading the Hostboot code from the PNOR, verifying its signature, then jumping into Hostboot code itself.

Note that the Bootloader code is part of the SBE SEEPROM contents.

Author: I've tried to keep this list as simple as possible, but now that it gave us some problems I think it is worth describing.

How exactly is SEEPROM updated with newer versions? I know this is done by  the Hostboot, so sometimes there is an additional reboot after flashing, but haven't found the code responsible for it yet (not that I've specifically looked for it). Are there any checks done? We were planning to put coreboot's bootblock in the place of HBBL, which would then parse PNOR and load the rest of stages, but we're not sure if this is feasible without modifications to the SBE code.

Reviewer 3: Probably over i2c? You can read/write the seeprom from the bmc (well, one of each pair in the cpu)
via the i2c sysfs interface; its presented as four logical at24-type eeproms (however the size in the
bmc devicetree is wrong so you have to bind it manually).

Reviewer 6: https://github.com/open-power/hostboot/blob/master/src/usr/sbe/sbe_update.C is the code that drives the SBE update from Hostboot.  

The SBE code doesn't care what it loads so you can freely replace HBBL.  However, you should be aware that the space is extremely limited.  I think we're literally down to having a handful of free bytes left.  If you need more space, that would require changes to the layout of the SBE code.

> You can read/write the seeprom from the bmc 
Note that the SEEPROM is typically not accessible from the BMC, unless you have security disabled with the jumper.

Author: Thanks @hanetzer, thanks @dcrowell77, both comments are extremely useful. Back to the drawing board I go :)

Reviewer 3: Would be interesting to turn the sbefw into coreboot as well, with the sbe runtime as its payload.

Author: I'm afraid coreboot is too big to fit there. Unless... Are you thinking about starting it on QEMU just for testing?

[Hostboot](https://git.raptorcs.com/git/talos-hostboot/) is the first component

> running on main cores. It is responsible mainly for the memory initialisation

Reviewer 5: ```suggestion
that runs on main cores. It is responsible mainly for the memory initialisation
```

> and training, it also initialises some of the platform buses.

Reviewer 5: ```suggestion
and training, it also starts some of the platform buses.
```

[Skiboot](https://git.raptorcs.com/git/talos-skiboot/) is then chainloaded. Its
tasks include initialisation of the rest of the hardware, e.g. PCI Express host
bus controllers. It also implements OPAL ([OpenPOWER Abstraction Layer](https://open-power.github.io/skiboot/doc/opal-spec.html#what-is-opal)),
which is an interface to access firmware services from an OS. This is similar to
the UEFI runtime services or legacy BIOS interrupt calls.

Next link in this chain is [Skiroot](https://github.com/open-power/linux) (a
small set of patches on top of the mainstream Linux kernel) accompanied by
[Petitboot](https://git.raptorcs.com/git/talos-petitboot/) (userspace
application). The latter is a bootloader, it uses kexec to load the final
operating system from any device that can be mounted by Linux, including network
boot.

There are also other pieces of firmware running on different chips. Those are
responsible for thermal and power management or CAPI (coherent accelerator

> processor interface) support. Most of them are loaded by the Hostboot.

Reviewer 2: @krystian-hebel I counted quite a lot of those components, so I would highlight how, despite being open, OpenPOWER Firmware ecosystem is, whet inevitably have to lead to issues during maintenance.

just the Hostboot part, at least for the time being. Earlier stages are run by a
separate hardware, and dividing Skiboot into "hardware" and "services" parts at
this stage may introduce additional bugs. It is safer to implement one component
at a time than fail trying to do them all in one go.

## Benefits

We can define two classes of possible benefits: one is what coreboot and its
community can gain, the other one is what end users get from it.

#### For users

Boot time reduction was one of the main reasons for doing this port. Hostboot
itself takes more than 1 minute to run. This is mostly spent on accessing flash
memory. It is split into multiple PNOR (Processor NOR, POWER name for flash)
partitions, summing up to just below 32 megabytes. It runs from cache until it
trains main memory, so it is limited to just 10 MB (size of L3 cache) split
between two cores. To make it possible to load and execute that amount of code,
[Hostboot uses on-demand paging](https://youtu.be/fTLsS_QZ8us?t=1559). It is
similar to swap used by operating systems, except that it uses ROM instead of
writable media, so only code and constant data can be discarded from memory (L3

> cache in this case).

Reviewer 6: This isn't true.  We have writable sections of the pnor, but they are only used to store a few writable persistent attributes and the VPD caches.

Another issue was nicely formulated by Timothy Pearson from Raptor CS:

> > Hostboot is a VM to run FSI routines, and the FSI routines are written one by

Reviewer: source?

> one by the hardware engineers designing the silicon itself.  This leads to
> overly complex and difficult to understand code, as you can see.

FSI is one of the buses Hostboot has to initialise. By simplifying the code we
can make it **easier to understand** (no need to translate virtual addresses to
the physical ones), **faster** (by discarding a dozen of function calls before
the proper write to the register happens) and **smaller**. Smaller code means
less time spent on reading from flash, but also more space for other components,
like new drivers for Skiroot.

#### For the coreboot community

coreboot would get a **new architecture** supported. Although there are stubs,
cross-compiler and libraries for PPC, the coreboot itself isn't in working state
when it comes to this architecture as of now. It would also get a **new, truly

> open** ([unless you want SATA controller](https://wiki.raptorcs.com/wiki/PM8068))

Reviewer: While we are at it, SSD (openssd-project.org/wiki/The_OpenSSD_Project) or Open GPU (https://www.crowdsupply.com/libre-risc-v/)

Author: Right, but IIUC this is for the controller on the disk side, not the one on the mainboard. For the second link I'm getting 404

platform.

PowerPC by default uses big endian. It is configurable at runtime, but the
libraries included in coreboot-sdk are compiled for BE. This gives a perfect
opportunity to validate that **coreboot works also for big endian platforms**.
By now we confirmed that there are parts in CBFS and FMAP which need small
fixes. The endianness of fields also needs to be properly documented for those

> structures, as they can be used both by a platform it is started on, as well as

Reviewer 5: It is started on/it is build on , nie daję suggest, bo nie wiem czy dobrze zrozumiałam kontekst, stąd prośba an zaminę it is

an application on a platform it is built on. Those two do not have to use the
same endianness.

The biggest achievement of this port would undoubtedly be bringing support for
**native DDR4 initialisation and training**. Even though the details of
accessing the hardware are platform-specific, the general process stays the
same. One caveat might be that Talos II supports only RDIMM (registered, or
buffered DIMM) and that is the only kind of memory sticks we are able to test,
however even if it won't work out of the box for different DIMM flavours, it

> should give a good base to build on.

Reviewer 5: give to whom? maybe it should be a good base?

Reviewer 2: @krystian-hebel benefit of that could be way better explained TBH this is a huge deal. I'm not sure what platforms operate on DDR with RDIMM, but we can imagine the situation that we get rid of FSP-M part or piece or PSP :thinking: and getting rid of memory initialization. Even if not x86 then RISC-V would benefit.

Reviewer: @KBanecka developers (newer DDR standards documentation without blobs or NDA, first, then end-users after base is worked on and others jump in code base.

## Current state of work

We have successfully started coreboot on the platform:

[![asciicast](https://asciinema.org/a/JQ1MaBSzGN1L1JcbgTX3G3kt6.svg)](https://asciinema.org/a/JQ1MaBSzGN1L1JcbgTX3G3kt6?speed=1)

> Actually, we were starting it much earlier than we knew it, because we used a

Reviewer 5: ...() than we got acknowledged with the platform?

Reviewer: Maybe being more precise on the coreboot stage that was successful here would be advisable then saying coreboot was started (as per tweets and everywhere else)

wrong address for serial port. It was caused mostly because of improper analysis
of Hostboot's code, so we ended up converting virtual address to a physical one
using a negative offset compared to what we should have used. This is what you
get from mixing virtual memory management and manual pointer calculations 🙂️

Since then we believe we have [mostly fixed all of the endianness issues](https://github.com/3mdeb/coreboot/commits/fmap_cbfs_endian),
so `Couldn't load romstage` became `Payload not loaded`, without any actual
initialisation code. After thorough tests and updates to the documentation we
will begin to upstream these changes.

If you think we can help in improving the security of your firmware or you are
looking for someone who can boost your product by leveraging advanced features
of used hardware platform, feel free to [book a call with us](https://calendly.com/3mdeb/consulting-remote-meeting)
or drop us email to `contact<at>3mdeb<dot>com`. If you are interested in similar

> content feel free to [sign up to our newsletter](http://eepurl.com/doF8GX)

Reviewer 2: @krystian-hebel I would improve this summary in the context of the article e.g. we would be glad to support the enabling of new  RISC-V or POWER10 platforms.

I'm missing the community excitement that we experienced. Maybe @miczyg1 can add something. There are also notes from OSFC which can be shared and reflected in this post.

We have to keep the bar high.

Reviewer: After the Power9, the Power10, then the power based laptop and finally an alternative to x86 to the actual exosystem by diversity of form factor choices, reduction of production costs. End user accessibility by cost factor. But first: openness for developers, better understanding of the architecture as a start.

I would try to find a way to point that yeah, actual prices of early adopters are high and require more radical people demanding a x86 alternative. But once the price of the boards go down, that CPUs become constructed not only for servers but for workstations and demand is there for laptop and lower power consumption, something else then the https://www.powerpc-notebook.org/en/ will be available. 

Something in the lines of: The Talos II is already RYF certified. It will now have coreboot, permitting payloads ecosystems to traverse x86 for performance and openness. Future ports to other boards will be eased. Costs of such boards will be reduced by demand.

Also.
By setting the bar higher for the coreboot supported platforms to finally have a real open source firmware existence (no PSP, no ME, no FSP, no e2dk) users and developers will finally have an option to choose from. Cloud providers will need to take a stance in terms of the platforms they choose in protecting their user's data, their privacy without choosing.

I do not know if we want to mix things and tackle the fact that QubesOS is taking small steps toward the server world (4.1 with headless remote support), that the network stack is taking necessary steps to be able to create DMZs (https://github.com/Rudd-O/qubes-network-server/), that Xen developers are ready to jump back into the PPC64 world, so that a real, strong alternative to x86 is showing the tip of its nose?

That the people actually choosing  virtualization platforms for their compatibility with kvm don't have to choose x86 anymore. That soon Xen will be part of the offer, and that now is the moment to show interest and support?

Not a marketing guy here as you see. But that is what I would love to read. Some hope getting of of x86 without having to recreate the wheel. Get that the ecosystem is transferred this time. So that other projects relying on coreboot can find their way into x86 alternatives without sacrificing performance/virtualization, nor choose between RAM availability. Privacy and blobless architecture is available now. The need for the common base for payload ecosystem will be done. And maybe even if bold, ask readers to say what they would love to see happen next!

Reviewer 2: @tlaurion 100% agree. @krystian-hebel any chance you can interleave Thierry comments into a summary?
